insights/core/spec_factory.py

```diff
@@ -719,6 +719,31 @@ def __call__(self, broker):
         raise ContentException("Can't list %s or nothing there." % p)
 
 
+class listglob(listdir):
+    """
+    List paths matching a glob pattern.
+
+    Args:
+        pattern (str): glob pattern to list.
+        context (ExecutionContext): the context under which the datasource
+            should run.
+        ignore (str): regular expression defining paths to ignore.
+
+    Returns:
+        function: A datasource that returns the list of paths that match
+            the given glob pattern.
+    """
+    def __call__(self, broker):
+        ctx = _get_context(self.context, broker)
+        p = os.path.join(ctx.root, self.path.lstrip('/'))
+        p = ctx.locate_path(p)
+        result = sorted(glob(p))
+        # generator expression; we don't need the full list at this step
+        result = (os.path.relpath(r, start=ctx.root) for r in result)
+        result = [r for r in result if not self.ignore_func(r)]
+        return result
```

Comment:
How about raising a ContentException when there is nothing matched, just like the `listdir`?

Reply:
I just saw your comment about this point in #3693, but I still recommend you to raise a ContentException here, since the empty `list` is not worth keeping and taking to the parser side. The final result for the parser would be the same: avoid invalid triggers (otherwise, you need to check empty by yourself). And the `ContentException` would be used for debugging/troubleshooting since it could be get from the `broker.exceptions`. Unless there is a rule that needs to check the empty result as a condition

Reply:
The rule that I need it for will hit if `listglob` returns an empty list. The rule will take additional measures to make sure the list is not empty by accident (e.g. a data collection error). 

Also note that there will be no `@parser` component involved. This datasource already provides structured data. Moreover, `@parser` components cannot work with datasources like `listdir` and `listglob`. The existing `listdir` specs are used as providers in `foreach_*` spec factories. Mine will be used by a "parser" implemented as `@combiner`.

As for the change that you are proposing:

The only case when there are little doubts about raising `ContentException` is when `listdir` is supposed to list a directory which doesn't exists.

In all other cases when `listdir` or `listglob` would return an empty list, the datasource does not know if it is due to the list being genuinely empty or due to a data collection error. There are two reasons for raising `ContentException` in this case:
1. To indicate the option of a data collection error, enforcing rules to take extra steps (specifying an optional dependency) before using the "empty result".
2. To optimize performance (based on the assumption that most applications will be interested in the **presence** of an item in the result).

I would argue that reason 1 is giving rule developers a false sense of security. If a rule is interested in the **absence** of an item in the result, the fact that `listdir` or `listglob` returned a non-empty list does not mean the list is complete. The list can be still incomplete due to data collection errors and the rule should use an independent channel to verify if is safe to use it (e.g. [gathers.json](https://github.com/openshift/insights-operator/blob/master/docs/insights-archive-sample/insights-operator/gathers.json) for Insights Operator archives.

I would argue that reason 2 is negligible. There might be a little overhead processing the components, but empty list means zero iterations of any processing code otherwise.

The `listdir` does not behave consistently like this and can return an empty list (see #3693). There are two options to align `listdir` and `listglob` behavior. Both options require `listdir` changes (I can prepare the changes; the changes do not break existing specs).

1. `listdir` does  not support globs; it raises `ContentException` when the specified path is not a directory. `listdir` and `listglob` can return an empty list.
2. `listdir` and `listglob` always raise `ContentException` instead of returning an empty list. (Whether `listdir` supports globs is not relevant in this case.)

I favor option 1; I think it provides a more flexible API and it matches my immediate use case better.

I can settle for option 2 if `listdir` is changed accordingly. I would still remove glob support from `listdir`.

WDYT?

Reply:
There's another option: add an `allow_empty` argument to the two spec factories. This would allow individual specs to declare if empty list is a valid result or not.

Reply:
We agreed separately on option 1. I'm closing this thread.